app/models/friendship.rb

```diff
@@ -1,4 +1,28 @@
 class Friendship < ApplicationRecord
+  # Callback
+
+  # Associations
   belongs_to :user
   belongs_to :friend, class_name: 'User'
-end
+
+  # Validations
+  validate :users_are_not_already_friends
+  validate :user_and_friend_are_not_the_same_person
+
+  private
+
+  def users_are_not_already_friends
```

Comment:
# :beers: 

Great work guys. We can also think about adding another layer of security here, at the database level. This isn't required but I encourage you to try it.

You see in your `schema.rb` you already have the ff line:

```ruby
t.index ["user_id", "friend_id"], name: "index_friendships_on_user_id_and_friend_id", unique: true
```

Which means if, for some reason, the user still manages to get past the model validation above, surely the database won't allow any duplicate record between the same `user_id` and `friend_id`. 

So if `ryto` has already added `mike` in the past (i.e, in the database we have a record where `user_id = ryto.id` and `friend_id = mike.id`) then `ryto` cannot anymore create another friendship with `mike`. 

But what if `mike` tries to create a duplicate friendship with `ryto` ? Your validation should stop that but at times it is possible that this kind of thing still goes through. We need to make sure the database knows about this.

What we can do is the ff:

- have a private method that will force a conflict of record:

```ruby
def force_record_conflict
  # 1. Simply return if there is already an existing friendship between user and friend
  # 2. If not then swap the user with the friend and the friend with the user
  self.user, self.friend = self.friend, self.user
end
```
What this will do is that if there is indeed an existing friendship between `ryto` and `mike`, no matter which of them adds the other, then the database won't allow a duplicate friendship between them to be created. Try it, let me know how it goes :)

Reply:
Hello @rvvergara
If I am no missing something, we have already handled this case by reusing `all_friends` method, that it will check for both side of friendship relations.
This is related spec:
https://github.com/mosaaleb/fakebook/blob/0ed33d5a62ed89d60dcd38f1cdc86d0783af2ca0/spec/models/request_spec.rb#L29-L38

Reply:
Hello @rvvergara, thanks for the continuous effort in trying to help us write better codes.

If we get you right, you are talking about [The Perils of Uniqueness Validations](https://thoughtbot.com/blog/the-perils-of-uniqueness-validations). Now, we are confused about a few things about your suggestion:

1. Referring to what we understand, database constraints are supposed to be reflected in the schema, but this suggestion is sounding similar to ordinary validation and not database level restriction.

2. After this method you're suggesting, we believe, we need a `before_create` callback or something similar to effect this method. Does that mean that callbacks are database level restrictions?

Coming to think of it, I think callbacks can also be called twice at the same time when two different users are trying to run the same process at the same time.

Finally, is the swapping you're suggesting not similar to the code below? We are ensuring that the action validates for `user_id, friend_id` and `friend_id, user_id`.

https://github.com/mosaaleb/fakebook/blob/0ed33d5a62ed89d60dcd38f1cdc86d0783af2ca0/app/models/friendship.rb#L14-L21

If there's something we are misunderstanding, please kindly shed some light. We hope to learn as much as possible from you before you zoom off on a job offer. :100: :100: :100:

Reply:
Hello @rvvergara, thanks for the continuous suggestions to help us write better codes.

We have implemented your suggestion, but differently. 

**Friendship database restriction**
https://github.com/mosaaleb/fakebook/blob/506aa4671cfd2ca12c14d7e6a810c8355f978501/db/schema.rb#L29-L38

**Request database restriction**
https://github.com/mosaaleb/fakebook/blob/506aa4671cfd2ca12c14d7e6a810c8355f978501/db/schema.rb#L72-L81

Both databases now work in accordance with your suggestions.

Reply:
Great job guys! I tried it and it works great!!

# :1st_place_medal: